waspc/src/Wasp/Node/Version.hs

```diff
@@ -16,26 +13,53 @@ import qualified Text.Regex.TDFA as R
 import qualified Wasp.SemanticVersion as SV
 import Wasp.Util (indent)
 
+-- | Wasp supports any node version equal or greater to this version.
+-- | We usually keep this one equal to the latest LTS.
+oldestWaspSupportedNodeVersion :: SV.Version
+oldestWaspSupportedNodeVersion = SV.Version 20 0 0
+
+data VersionCheckResult
+  = VersionCheckFail !ErrorMessage
+  | VersionCheckSuccess !(Maybe WarningMessage) !SV.Version
+
+type WarningMessage = String
+
+type ErrorMessage = String
+
 -- | Gets the installed node version, if any is installed, and checks that it
 -- meets Wasp's version requirement.
---
--- Returns a string representing the error
--- condition if node's version could not be found or if the version does not
--- meet the requirements.
-getAndCheckNodeVersion :: IO (Either String SV.Version)
+getAndCheckNodeVersion :: IO VersionCheckResult
 getAndCheckNodeVersion =
   getNodeVersion >>= \case
-    Left errorMsg -> return $ Left errorMsg
-    Right nodeVersion ->
-      if SV.isVersionInRange nodeVersion nodeVersionRange
-        then return $ Right nodeVersion
-        else return $ Left $ makeNodeVersionMismatchMessage nodeVersion
+    Left errorMsg -> return $ VersionCheckFail errorMsg
+    Right userNodeVersion ->
+      return $
+        if SV.isVersionInRange userNodeVersion $ SV.Range [SV.gte oldestWaspSupportedNodeVersion]
+          then VersionCheckSuccess Nothing userNodeVersion
+          else VersionCheckFail $ makeNodeVersionMismatchMessage userNodeVersion
+
+makeNodeVersionMismatchMessage :: SV.Version -> String
+makeNodeVersionMismatchMessage nodeVersion =
+  unlines
+    [ unwords
+        [ "Your Node version does not meet Wasp's requirements!",
+          "You are running Node " ++ show nodeVersion ++ "."
+        ],
+      waspNodeRequirementMessage
+    ]
+
+waspNodeRequirementMessage :: String
+waspNodeRequirementMessage =
+  unwords
+    [ "Wasp requires that you use at least Node version " ++ show oldestWaspSupportedNodeVersion ++ " or higher. Also, node binary must be available in the PATH.",
```

Comment:
Ok, I both simplified the message and made sure that I talk about version mismatch only when there is version mismatch, and about PATH only when PATH is potentially the problem.

Reply:
I think this is still a little too verbose and too cluttered (not enough spacing).
I have a feeling our users will still have problems parsing the output.

What are your concerns with my suggestion above?

> [I] made sure that I talk about version mismatch only when there is version mismatch, and **about PATH only when PATH is potentially the problem**.

This seems to be broken:

![image](https://github.com/wasp-lang/wasp/assets/22752307/fad4684a-7d7d-4922-bde4-34eafc41545a)

Reply:
I'll analyze both messages separately.

I'm only going into so much detail because the old version of this message was the biggest stumbling point on our Feedback Thursdays. It's also something many of our users see.
Therefore, I think it makes sense to spend time here and make it as clear as possible.

### When `node` isn't in the `PATH`
This is how the full message looks now:
```
❌ --- [Error] Node requirement not met: ------------------------------------------

Running `node --version` failed.
  `node` command not found!
Make sure you have node installed and in your PATH.

```
This is how I imagine the ideal message (I left the banner out on purpose, but keep reading :)):
```
node command not found.
Make sure you have node installed and in your PATH.
```
Explanation:
- Our users don't need to know how we figured out `node` is missing. The `node --version` part doesn't tell them anything important and can confuse them.
- I think backticks are unusual in the terminal, but could be misremembering.
- I've removed the banner because I seem to remember it drawing the users' attention away from the important part of the message (on Feedback Thursdays). That said, I'm aware we're probably keeping it.

If we're keeping the banner, then I like this:
```
❌ --- [Error] Node requirement not met: ------------------------------------------

node command not found!
Make sure you have node installed and in your PATH.
```

### When `node` is in the `PATH`

This is how the full message looks now:
```
❌ --- [Error] Node requirement not met: ------------------------------------------

Your Node version does not meet Wasp's requirements! You are running Node 16.18.0 .
Wasp requires Node version 18.0.0 or higher.

```

Specific things I would change:
- _Your Node version does not meet Wasp's requirements!_ - This part tells them the same thing the title already told them. It doesn't convey any new information and is the longest sentence in the output.
- _Wasp requires Node version 18.0.0 or higher._ - I believe it's better if we addressed the user directly.
- The ordering of messages - I'd state our requirements before telling them why it doesn't work for them.
- The banner - Reasons are explained above, but I realize it probably stays.

With all that in mind, this is how I imagine the ideal message (again without the banner):
```
To use Wasp, you need Node v18.0.0 or higher.
You are running Node 16.18.0 .
```

It conveys the same message with fewer words and simpler language, and gets to the point immediately: We first tell them what we expect, and then we tell them how their system fails said expectation.

I realize that you probably want to/have to keep the banner, which makes my propsed message look like this:
```
❌ --- [Error] Node requirement not met: ------------------------------------------

To use Wasp, you need Node v18.0.0 or higher.
You are running Node 16.18.0.
```

Reply:
Ok got it, so the main things are:
1. When node not in PATH, don't mentioned `node --version`.
2. When wrong node version, make message shorter and first mention requirements, then what they have.

While lot of them were stumbling on this message, as you said, from what I got from the feedback, reason wasn't them not being able to understand the message, but instead being confused as to why they have to install another node version. Or, they just didn't know what PATH is. So all of these suggestions, while possibly (or maybe not) an improvement, don't matter much I believe.

To quickly address some of them:
1. I love seeing which command went wrong, because then I can try to run it on my own. So if `wasp` tells me it failed in running `node --version`, then I can try to run that on my own and see what happens. So I put that on purpose. Also, we make sure we are not giving them wrong info: maybe we are telling them that we can't find `node`, but that is not correct, we could but running `node --version` failed for some reason (e.g. permissions). I have no idea if other people will find this useful or confusing, we can only guess.
2. I usually don't read the banner, just the error message, so I wanted to repeat the information in the message, to make it very clear what is wrong and not have them wonder if they haven't seen the banner.

Printing Wasp version before Node version -> kind of makes sense, maybe, in theory, but sounds irrelevant in practice to me.

Getting rid of the banner -> yeah, it is not so easy to get rid of it at the moment.

Conclusion: thanks for the detailed write up above, you did a nice analysis there and it all makes sense to me, but also what I did so far makes sense to me, and I don't have a good way to choose one over the other -> also, I don't think it matters, I think we solved the important part. I had to ask for clarification though because I didn't know if I was missing something more crucial. But I would wrap up with this then, and let's see how people will react to it, we can further improve it if needed.